client/consensus/grandpa/authorities.go

```diff
@@ -160,51 +231,414 @@ func (authSet *AuthoritySet) addForcedChange(pending PendingChange, isDescendent
+	roots := authSet.pendingStandardChanges.Roots()
+	for i := 0; i < len(roots); i++ {
+		effectiveNumber := roots[i].change.EffectiveNumber()
+		if effectiveNumber >= min {
+			if limit == nil {
+				limit = &effectiveNumber
+			} else if effectiveNumber < *limit {
+				*limit = effectiveNumber
+			}
+		}
+	}
+	return limit
 }
 
-// SharedAuthoritySet A shared authority set.
-// TODO implement shared logic
-type SharedAuthoritySet struct {
-	authoritySet AuthoritySet
+// ApplyForcedChanges Apply or prune any pending transitions based on a best-block trigger.
+//
+// Returns a pointer to the median and new_set when a forced change has occurred. The
+// median represents the median last finalized block at the time the change
+// was signaled, and it should be used as the canon block when starting the
+// new grandpa voter. Only alters the internal state in this case.
+//
+// These transitions are always forced and do not lead to justifications
+// which light clients can follow.
+//
+// Forced changes can only be applied after all pending standard changes
+// that it depends on have been applied. If any pending standard change
+// exists that is an ancestor of a given forced changed and which effective
+// block number is lower than the last finalized block (as defined by the
+// forced change), then the forced change cannot be applied. An error will
+// be returned in that case which will prevent block import.
+func (authSet *AuthoritySet) applyForcedChanges(bestHash common.Hash,
+	bestNumber uint,
+	isDescendentOf IsDescendentOf,
+	initialSync bool,
+	telemetry *telemetry.Client) (newSet *appliedChanges, err error) {
+
+	for _, change := range authSet.pendingForcedChanges {
+		effectiveNumber := change.EffectiveNumber()
+		if effectiveNumber > bestNumber {
+			continue
+		} else if effectiveNumber == bestNumber {
+			// check if the given best block is in the same branch as
+			// the block that signaled the change.
+			isDesc, err := isDescendentOf(change.canonHash, bestHash)
+			if err != nil {
+				return nil, err
+			}
+			if change.canonHash == bestHash || isDesc {
+				medianLastFinalized := change.delayKind.value.(Best).medianLastFinalized
+
+				roots := authSet.pendingStandardChanges.Roots()
+				for _, standardChangeNode := range roots {
+					standardChange := standardChangeNode.change
+
+					isDescStandard, err := isDescendentOf(standardChange.canonHash, change.canonHash)
+					if err != nil {
+						return nil, err
+					}
+					if standardChange.EffectiveNumber() <= medianLastFinalized && isDescStandard {
+						logger.Infof(
+							"Not applying authority set change forced at block %d, due to pending standard change at block %d",
+							change.canonHeight, standardChange.EffectiveNumber())
+						return nil, errForcedAuthoritySetChangeDependencyUnsatisfied
+					}
+				}
+
+				// apply this change: make the set canonical
+				logger.Infof("%v 👴 Applying authority set change forced at block %d", initialSync, change.canonHeight)
```

Comment:
So actually I realize that it is used as a conditional variable to avoid printing logs during bootstrap. Do you think this would be nice to have? Or better to just either print it or not print it?

Reply:
I think this could be printed for Bootstrap sync as well! And I would suggest using the `#` when printing numbers so its easy to identify it in the log lines

Reply:
updated!